server/broadcast.go

```diff
@@ -376,10 +377,12 @@ func transcodeSegment(cxn *rtmpConnection, seg *stream.HLSSegment, name string,
 	}
 	res, err := SubmitSegment(sess, seg, nonce)
 	if err != nil || res == nil {
+		glog.Error("unable to submit segment err=", err)
```

Comment:
Except for a few cases, most of the errors encountered in `SubmitSegment()` are logged in place so adding additional log statements here would lead to two log statements for most errors (and the second one has less information than the first one). So, I don't think we should add an additional log statement here right now. The simplest alternative that can make sure all errors are logged might be to add in place log statements for the remaining cases in `SubmitSegment()` - one example is the error from [newBalanceUpdate()](https://github.com/livepeer/go-livepeer/blob/ce819a15a616277a7081b5cb91339c4121a22538/server/segment_rpc.go#L302).

Reply:
Ah my bad this is actually not necessary. We already log the error when sender doesn't have enough balance.

We just never reach that code path if the sender doesn't have deposit & reserve because instead we'll already get an error because the max face value is 0 when reserve is 0. 

`E0325 13:30:42.773116   85004 broadcast.go:364] Invalid ticket params err=ticket faceValue 10000000000000000 > max faceValue 0`

Previously we would only validate the ticket params _after_ self-validating as a sender. 

I don't deem additional error logs in SubmitSegment necessary at this point for those cases that currently don't have error logging. WDYT?

Reply:
We could add a validateSender check https://github.com/livepeer/go-livepeer/blob/ce819a15a616277a7081b5cb91339c4121a22538/server/broadcast.go#L362 for a more informative error when the sender doesn't have enough balance.

Reply:
I'm on board with returning a more informative error from `ValidateTicketParams()`. Maybe something like this?

```golang
fmt.Errorf("faceValue too high faceValue=%v maxFaceValue=%v deposit=%v, ticketParams.FaceValue, maxFaceValue, sender.Deposit)
```

Reply:
I'm on board with making the error clearer. One thought is that the zero deposit check could be added to `validateSender()` and we could call `validateSender()` in `ValidateTicketParams()` so that the check could be run whenever we're preparing to submit a segment.

Reply:
Ok sorry I didn't look at the entire context for this.

I thought we check for 0 deposit/reserve in `validateSender` but this is not the case , what we do check is that the broadcasting funds are not set to unlock soon - https://github.com/livepeer/go-livepeer/pull/1318/files

When we run into a `validateSender` error we close the RTMP connection. 

Currently when the sender has 0 deposit/reserve we run into a `ticket faceVal > max faceVal` error emitted by  `ValidateTicketParams` and never reach `validateSender` in `SubmitSegment`  this error does _NOT_ close the RTMP connection and does not necessarily indicate 0 broadcasting funds. 

I think a  case  where the broadcaster doesn't have broadcasting funds  _SHOULD_ close the RTMP connection however. So how about we remove sender validation from `CreateTicketBatch` and instead export it and call `ValidateSender` in `transcodeSegment`.

Reply:
Can we just call `validateSender()` in `ValidateTicketParams()`? Seems like it accomplishes the same thing if `validateSender()` contains a zero deposit check. Would be good to have a zero reserve check in there as well.

Reply:
Ah yes, that's true. I would only check the reserve, as long as there is reserve the broadcaster does have "broadcasting funds" in theory and can still transcode potentially even when the deposit is 0 so that shouldn't lead to closing the RTMP connection.

Reply:
The max faceValue check in `validateTicketParams()` will fail if the deposit is zero so B still would not send out a stream in that case. Might be worth adding the zero deposit check in `validateSender()` just to return a clear and explicit error.

Reply:
fixed in 571ab3d

Reply:
Doesn't look like the check mentioned in this thread is in 571ab3d - is the plan to add that check in this PR?

Reply:
I believe with the current changes we don't need to add additional error logging in SubmitSegment, we now cover all critical user-facing error logs.

Reply:
I'm referring to the deposit check in `validateSender()` that is discussed in the above comments. Wasn't sure if the intent here was to include it in this PR. FWIW I'm fine with not including that in this PR.

Reply:
Ah looks I forgot to push up the commit dae2a45